src/abstract.jl

```diff
@@ -0,0 +1,36 @@
+"""
+    AbstractADType
+
+Supertype for all AD choices.
+"""
+abstract type AbstractADType{mode} end
+
+"""
+    AbstractFiniteDifferencesMode
+
+Supertype for AD choices that rely on [finite differences](https://en.wikipedia.org/wiki/Finite_difference).
+"""
+const AbstractFiniteDifferencesMode = AbstractADType{:finite}
```

Comment:
Why symbols here? Normally we would use an EnumX to stop comparison issues with spelling errors. But I don't quite get what this symbol is trying to do, is it supposed to specify a trait? Why not define a trait function then?

Reply:
Or rather, why is this not actually an abstract type?

Reply:
The idea was that `AutoSparse(AutoForwardDiff())` should have the same supertype as `AutoForwardDiff()`, aka `AbstractForwardMode`. But it is impossible to do the following in Julia (see this [Slack exchange](https://julialang.slack.com/archives/C6A044SQH/p1712899380312649))
```julia
struct AutoSparse{AD} <: TheSpecificSuperTypeOfAD <: AbstractADType end
```
So the trick I found is to make `AbstractADType` parametric, and to redefine `AbstractForwardMode` and its friends as aliases.
I don't really care if we use `Symbol`s or an `Enum` here, I can change to an `Enum` if you think it's better. But that is the rationale anyway.

Reply:
What's the reason for not relying on trait functions for this information instead?

Reply:
If we want to be more ambitious (but more breakin) we can replace it all with a `mode` trait, which is less cumbersome than this subtyping trick. It might make sense for backends that are both forward- and reverse-compatible, like Enzyme. Internally, it is what DifferentiationInterface does anyway.

Reply:
The reason for not using traits is that I wanted to do a soft PR that breaks as little as possible. But as you can see below, it is the nicer option, especially for backends who can do both forward and reverse mode:

https://github.com/gdalle/DifferentiationInterface.jl/blob/84d98a356a7e963b4b1acec09904edc78a70a033/DifferentiationInterface/src/traits.jl#L17-L20

https://github.com/gdalle/DifferentiationInterface.jl/blob/84d98a356a7e963b4b1acec09904edc78a70a033/DifferentiationInterface/ext/DifferentiationInterfaceEnzymeExt/DifferentiationInterfaceEnzymeExt.jl#L47-L48

Reply:
I think a mode trait would be much more desirable, indeed since something like Enzyme can have two modes and thus you'd need to split it to two types if you're to do it like this.

Reply:
Yeah it's rather uncomfortable to have it be a subtype of neither forward nor reverse mode. I'll update the PR with a trait mechanism instead, but it's gonna be more breaking as a result